mailbox/api/src/main/java/org/apache/james/mailbox/MessageManager.java

```diff
@@ -190,6 +190,15 @@
     MetaData getMetaData(boolean resetRecent, MailboxSession mailboxSession, MessageManager.MetaData.FetchGroup fetchGroup) throws MailboxException;
 
     /**
+     * Offers direct access to UIDs. This allows per implementation optimisation for retrieving the whole UIDs of a mailbox.
+     *
+     * @param session MailboxSession for the user performing the request.
+     * @return Iterator of MessageUids in this mailbox. They are not necessarily sorted.
+     * @throws MailboxException
+     */
+    Iterator<MessageUid> getUids(MailboxSession session) throws MailboxException;
```

Comment:
I don't think having "two ways to retrieve" the information is necessary bad. If one of those way do it differently than the other way and that it is clear why not. 
The case I have in mind is when a way X is a specialization of the API used by a way Y. So X can be optimized by this way.